### PR TITLE
[Terrarium - Providers]: Implementation of gRPC backend microservice integrated with DynamoDB for Data Persistence

### PR DESCRIPTION
**Pull request by anjanikshree12**
This PR includes:

- protobuff definitions for provider, version manager service
- Implementation of Version Manager Service with following functionalities:
   - To List Provider Versions (to be used by protocol apis for terraform)
   - To Get Provider's specific version's metadata (to be used by protocol apis for terraform)
   - To List Providers registered with the system (to be used by web service for UI)
   - To Get a specific provider's data (to be used by web service for UI)
   - To Register a provider with the system 
   - To Publish a version (only versions that have been published should be reported to other services)
   - To Deregister a provider (using providerID & it's version) from the system.
-    Unit tests added for functions (and tested with all unit tests passing)
-    Introduced a new package for common gateway service to both modules & providers
-    Introduced a new package for common service to both modules & providers

## Testing Details:
- Metadata response as expected by the provider registry protocol:
image
image

- Metadata reponse required by REST handler to be served to the UI:
image
image

- UI:
image
image

### CODE REVIEW

File: internal/provider/services/service.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package services
+
+import (
+	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+type Service interface {
+	RegisterWithServer(grpcServer grpc.ServiceRegistrar) error
```
Comment (adcharre) [inline internal/provider/services/service.go:10]: As mentioned in another comment, lets move this definition up to a package (create one if needs be) common to both Module and Provider

File: internal/provider/services/gateway/gateway.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	terminateProviderVersion := services.TerminateVersionRequest{
+		Provider: request.GetProvider(),
+	}
+
+	publishRequest := services.TerminateVersionRequest{
+		Provider: request.GetProvider(),
+	}
+
+	span := trace.SpanFromContext(ctx)
+
```
Comment (adcharre) [inline internal/provider/services/gateway/gateway.go:80]: You should set the span attributes here (like in the `RegisterWithClient` function above) for the request to this function.

File: internal/provider/services/gateway/gateway.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	if request.GetAction() == terrarium.EndProviderRequest_DISCARD_VERSION {
+		log.Println("Abort Provider Version => Version Manager")
+		if res, delegateError := client.AbortProviderVersion(ctx, &terminateProviderVersion); delegateError != nil {
+			log.Printf("Failed: %v", delegateError)
+			span.RecordError(delegateError)
+			return nil, delegateError
+		} else {
+			log.Println("Done <= Version Manager")
+			span.AddEvent("Successfully aborted provider version", trace.WithAttributes(attribute.String("Provider Name", request.Provider.GetName()), attribute.String("Provider Version", request.Provider.GetVersion())))
```
Comment (adcharre) [inline internal/provider/services/gateway/gateway.go:89]: Don't need this if you specify the span attributes. Also this event doesn't really add any value - the usual case would be for the operation to succeed, therefore the more interesting case is if something goes wrong where you're recording the error.

File: internal/provider/services/gateway/gateway.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+		}
+	} else if request.GetAction() == terrarium.EndProviderRequest_PUBLISH {
+		log.Println("Pubish Provider => Version Manager")
+		if res, delegateError := client.PublishVersion(ctx, &publishRequest); delegateError != nil {
+			log.Printf("Failed: %v", delegateError)
+			span.RecordError(delegateError)
+			return nil, delegateError
+		} else {
+			log.Println("Done <= Version Manager")
+			span.AddEvent("Successfully published version", trace.WithAttributes(attribute.String("Provider Name", request.Provider.GetName()), attribute.String("Provider Version", request.Provider.GetVersion())))
```
Comment (adcharre) [inline internal/provider/services/gateway/gateway.go:100]: Same as above no need for this.

File: cmd/allInOne.go [hunk 5]
```diff
@@ -137,9 +151,10 @@ func init() {
 	allInOneCmd.Flags().StringVar(&registrar.RegistrarTableName, "registrar-table", registrar.DefaultRegistrarTableName, "Module Registrar table name")
 	allInOneCmd.Flags().StringVar(&dependency_manager.ModuleDependenciesTableName, "module-dependencies-table", dependency_manager.DefaultModuleDependenciesTableName, "Module dependencies table name")
 	allInOneCmd.Flags().StringVar(&dependency_manager.ContainerDependenciesTableName, "container-dependencies-table", dependency_manager.DefaultContainerDependenciesTableName, "Module container dependencies table name")
+	allInOneCmd.Flags().StringVar(&providerVersionManager.VersionsTableName, "provider-table", providerVersionManager.DefaultProviderVersionsTableName, "Provider Version Manager table name")
 }
 
-func startAllInOneGrpcServices(services []services2.Service, endpoint string) {
+func startAllInOneGrpcServices(services []services2.Service, providerServices []providerServices.Service, endpoint string) {
```
[the remark this answers is not in the document]
Comment (anjanikshree12) [reply cmd/allInOne.go]: Introduced a new package and updated the both services to use this.

File: internal/provider/services/gateway/gateway.go (hunk 2 above)
Comment (anjanikshree12) [reply internal/provider/services/gateway/gateway.go]: Added.

File: internal/provider/services/gateway/gateway.go (hunk 3 above)
Comment (anjanikshree12) [reply internal/provider/services/gateway/gateway.go]: removed span events.

File: internal/provider/services/service.go (hunk 1 above)
Comment (anjanikshree12) [reply internal/provider/services/service.go]: yes, created a new package for gateway and service which is now common to both modules & providers.

File: internal/provider/services/version_manager/version_manager.go [hunk 6]
```diff
@@ -0,0 +1,596 @@
...
+
+	log.Println("New provider registered.")
+	return ProviderRegistered, nil
+}
+
+// ListProviderVersions Retrieve all versions of a given provider and return an array of versionItems.
+// Only versions that have been published should be reported.
+func (s *VersionManagerService) ListProviderVersions(ctx context.Context, request *services.ProviderName) (*services.ProviderVersionsResponse, error) {
+
+	filter := expression.And(
```
Comment (adcharre) [inline internal/provider/services/version_manager/version_manager.go:295]: Missing OpenTelemetry span

File: internal/provider/services/version_manager/version_manager.go [hunk 7]
```diff
@@ -0,0 +1,596 @@
...
+		}
+	}
+	grpcResponse.Versions = sortedVersions
+
+	return &grpcResponse, nil
+}
+
+func (s *VersionManagerService) GetVersionData(ctx context.Context, request *services.VersionDataRequest) (*services.PlatformMetadataResponse, error) {
+
+	providerID, err := attributevalue.Marshal(request.GetName())
```
Comment (adcharre) [inline internal/provider/services/version_manager/version_manager.go:363]: Missing OpenTelemetry span

File: internal/provider/services/version_manager/version_manager.go [hunk 8]
```diff
@@ -0,0 +1,596 @@
...
+			return nil, MarshalProviderError
+		}
+		return providerMetadata, nil
+	}
+}
+
+func (s *VersionManagerService) ListProviders(ctx context.Context, request *services.ListProvidersRequest) (*services.ListProvidersResponse, error) {
+
+	// Initialize a map to store providers uniquely
+	uniqueProviders := make(map[string]*services.ListProviderItem)
```
Comment (adcharre) [inline internal/provider/services/version_manager/version_manager.go:398]: Missing OpenTelemetry span

File: internal/provider/services/version_manager/version_manager.go [hunk 9]
```diff
@@ -0,0 +1,596 @@
...
+	grpcResponse := services.ListProvidersResponse{
+		Providers: providersList,
+	}
+
+	return &grpcResponse, nil
+}
+
+func (s *VersionManagerService) GetProvider(ctx context.Context, request *services.ProviderName) (*services.GetProviderResponse, error) {
+
+	filter := expression.Name("name").Equal(expression.Value(request.GetProvider()))
```
Comment (adcharre) [inline internal/provider/services/version_manager/version_manager.go:449]: Missing OpenTelemetry span

File: internal/provider/services/version_manager/version_manager.go (hunk 6 above)
Comment (anjanikshree12) [reply internal/provider/services/version_manager/version_manager.go:300]: Added missing opentelemetry spans

**Review by adcharre — COMMENTED** [review]
Very close now - just need to add the `defer span.End()` to make sure all spans are closed off.

File: internal/provider/services/version_manager/version_manager.go [hunk 10]
```diff
@@ -0,0 +1,632 @@
...
+		"name":    providerName,
+		"version": providerVersion,
+	}, nil
+}
+
+// AbortProviderVersion removes a Version of a Provider.
+func (s *VersionManagerService) AbortProviderVersion(ctx context.Context, request *services.TerminateVersionRequest) (*terrarium.Response, error) {
+	log.Println("Aborting provider version.")
+
+	span := trace.SpanFromContext(ctx)
```
[the remark this answers is not in the document]
Comment (adcharre) [reply internal/provider/services/version_manager/version_manager.go:105]: Having looked at this again - I'm not convinced that we do need the span.End() - can you check and confirm it's not required here?

**Review by adcharre — APPROVED** [review]

File: internal/provider/services/version_manager/version_manager.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (anjanikshree12) [reply internal/provider/services/version_manager/version_manager.go:105]: I believe that if the span is created by a tracer then it needs to be end.
https://opentelemetry.io/docs/languages/go/instrumentation/#creating-spans
but if we are pulling the span from context, it's optional to use span.end to close the span.
https://opentelemetry.io/docs/languages/go/instrumentation/#get-the-current-span

**Pull request merged**